ambry-router/src/main/java/com/github/ambry/router/OperationController.java

```diff
@@ -310,14 +313,19 @@ protected void doOperationTowardsMaybeCompositeBlob(final String blobIdStr, Futu
               String.format("GET blob call returned the blob instead of just the store keys (before {} )",
                   helper.getOpName()), RouterErrorCode.UnexpectedInternalError);
           NonBlockingRouter.completeOperation(futureResult, callback, null, exception, false);
+        } else if (helper.getAlreadyUpdated().apply(result.getBlobInfo()) && canRelyOnMetadataForUpdate(
+            result.getBlobInfo())) {
+          // If we are here then we can rely that the metadata chunk was updated only after all data chunks were updated.
+          // So this means we are sure that all data chunks are successfully updated.
+          // Send an update for metadata chunk only to ensure that metadata chunk's update meets quorum.
+          helper.getDoOperation().accept(blobIdStr, Collections.emptyList());
```

Comment:
If we are here, does it mean that metadata blob (`blobIdStr`) has been TTL updated? If so, may be we can just complete the router operation here? Seems like above would try to update TTL again on metadata blob?

Reply:
If we are here, the only claim we can make is that all chunkIds have been ttl updated successfully in a previously issues ttl update (in the sense that all ttl updates for the chunks got the quorum). However, it could still have happened that metadata chunk ttl update didn't get quorum, and the overall ttl update had failed last time due to that because the Get operation to get the metadata chunk status looks at only one replica. Hence, in order to ensure that metadata chunk also got quorum, the easiest thing to do is to issue ttl update for metadata chunk.